### PR TITLE
SSRF helper validates a hostname then lets httpx re-resolve it (tsk-6uymvv)

### PR DESCRIPTION
**Pull request by jaylfc**
CARD TITLE (intent, not commit subject): SSRF helper validates a hostname then lets httpx re-resolve it - a low-TTL nameserver can answer public to the check and 127.0.0.1 to the connection
Autonomous build of board card tsk-6uymvv.

## What changed

`validate_url_or_raise` resolved a hostname, checked every answer, and then
returned nothing but permission. The caller's `httpx` client resolved the same
name a second time when it opened the connection, so a nameserver the attacker
controls could answer public to the check and `127.0.0.1` to the connection with
every check in the module still passing. Adding blocklist entries cannot fix
that; the address that was checked has to be the address connected to.

**The pin.** `tinyagentos/routes/desktop_browser/ssrf.py` gains
`guarded_async_client(...)`: an `httpx.AsyncClient` whose connection pool uses a
`_PinnedResolutionBackend`. `httpcore` calls that backend's `connect_tcp` with
the hostname from the request URL - the exact place the second lookup used to
happen - and the backend resolves it, runs the blocklist over the answer, and
hands the socket layer the literal address instead of the name. One lookup, and
it is the checked one.

This is the card's option 2 (pin at the transport), but at the network backend
rather than by rewriting the request URL to the IP. Rewriting the URL would have
worked, and it is what the card sketches, but it leaks the literal upward:
`respx`-based tests, redirect resolution, cookie scoping, and `response.url` all
read the request URL. Pinning below it leaves all of that stock `httpx`, and
means **TLS needs no special handling at all** - `httpcore` still derives SNI and
the certificate-verification hostname from the original name, and `verify` stays
on (test: `test_tls_verification_stays_on_and_the_url_keeps_the_hostname`).
The transport raises at construction if a future `httpx` moves that seam, rather
than quietly handing back an unpinned client.

**Every call site, not one caller.** All eight `validate_url_or_raise` call sites
were enumerated; the six that then fetch now fetch through a guarded client:

| call site | change |
| --- | --- |
| `routes/desktop_browser/proxy.py` | `httpx.AsyncClient` -> `guarded_async_client` |
| `routes/desktop_browser/extract.py` | same |
| `routes/desktop_browser/download.py` | same |
| `library_pipeline.py` (`WebProcessor`, the card's representative caller) | same |
| `peer.py` (`deliver_handshake`, own-client branch) | same |
| `push/unifiedpush.py` (`HttpUnifiedPushSender` default client) | same, `allow_private=True` to mirror its own `send()` validation |
| `knowledge_ingest.py` (`_download_article`) | new `fetch_client=` injection, defaulting to a fresh guarded client per download |
| `routes/devices.py` | untouched - it validates a push token at registration and never fetches, so there is nothing to pin |

`knowledge_ingest` needed the extra seam because its injected `http_client` is
the app-wide one, also used to reach the LLM backend and qmd on loopback:
guarding *that* client would have blocked our own services. Trusted-internal and
untrusted-outbound are now separate clients, which is the boundary that was
missing.

`validate_url_or_raise` also stops returning `None` - it returns the addresses it
approved (order-preserving; the resolver's RFC 6724 order was previously
destroyed by a `set`). The early call at each site is kept as a fail-fast with a
readable reason for the 403; the transport is the enforcement.

**Scope refutation.** PR #2070 was not touched: as the card says, it uses the
helper correctly, and this defect is not in its code.

## RED FIRST (pasted)

Run at the base ref (`origin/dev`), with only the new test file added:

```
$ .venv/bin/python -m pytest tests/test_ssrf_rebinding.py -q -p no:cacheprovider; echo "exit $?"
F.                                                                       [100%]
=================================== FAILURES ===================================
__________________ test_second_lookup_to_loopback_is_refused ___________________

lib_store = <tinyagentos.library_store.LibraryStore object at 0x73a07c8957f0>
storage_dir = PosixPath('/tmp/tmp3cxh7mmw')
page_servers = <test_ssrf_rebinding._ConnectionRecorder object at 0x73a07c895e80>

    @pytest.mark.asyncio
    async def test_second_lookup_to_loopback_is_refused(lib_store, storage_dir, page_servers):
        """A nameserver that answers public, then loopback, must not be followed."""
        resolver = _ScriptedResolver({"rebind.test": [_PUBLIC, _INTERNAL]})

        blocked: SsrfBlockedError | None = None
        with patch("socket.getaddrinfo", resolver), page_servers.patched():
            try:
                await _fetch_through_web_processor(
                    lib_store, storage_dir, "http://rebind.test/page",
                )
            except SsrfBlockedError as e:
                blocked = e

    reached = page_servers.connected[-1] if page_servers.connected else "(no connection)"
>       assert blocked is not None, (
            f"expected SsrfBlockedError, but the fetch reached {reached}"
        )
E       AssertionError: expected SsrfBlockedError, but the fetch reached 127.0.0.1
E       assert None is not None

tests/test_ssrf_rebinding.py:213: AssertionError
=========================== short test summary item ============================
FAILED tests/test_ssrf_rebinding.py::test_second_lookup_to_loopback_is_refused
1 failed, 1 passed in 4.73s
exit 1
```

The failing test drives a real caller (`WebProcessor.process`) through a scripted
`socket.getaddrinfo` that answers a public address on the first call for
`rebind.test` and `127.0.0.1` on the second. The connection is intercepted at
`httpcore`'s network backend - the place a real connection resolves and opens the
socket - which performs that second lookup for real and then opens a real socket
to a loopback stand-in for the internal service. On `dev` the fetch reaches it.

`test_agreeing_lookups_still_fetch` is the control that passes in the same run: a
hostname whose two lookups agree on a public address still fetches, and asserts
the fetched text is the public page. Without it, breaking all outbound fetching
would turn the red test green.

## GREEN

```
$ .venv/bin/python -m pytest tests/test_ssrf_rebinding.py -q -p no:cacheprovider; echo "exit $?"
..                                                                       [100%]
2 passed in 2.16s
exit 0
```

Affected modules, at the pre-rebase head:

```
$ .venv/bin/python -m pytest tests/routes/desktop_browser tests/test_library.py \
    tests/test_knowledge_ingest.py tests/push/test_unifiedpush.py tests/test_contacts_peer.py \
    -q -p no:cacheprovider
746 passed, 27 warnings in 1513.26s (0:25:13)
```

Re-run after rebasing onto current `origin/dev`:

```
$ .venv/bin/python -m pytest tests/test_ssrf_rebinding.py tests/routes/desktop_browser/test_ssrf.py \
    tests/test_knowledge_ingest.py tests/push/test_unifiedpush.py -q -p no:cacheprovider; echo "exit $?"
85 passed, 1 warning in 1.86s
exit 0
```

`tests/routes/desktop_browser/test_ssrf.py` gains seven unit tests for the pin
itself: the validator returns the addresses it approved in resolver order, the
guarded client really installs `_PinnedResolutionBackend`, `connect_tcp` hands
the socket the checked literal, a blocked answer is refused before the inner
backend is touched, `allow_private` reaches the pin, TLS verification stays on
with the hostname intact, and unix-socket connections are refused.

## Docs

`python3 scripts/check_doc_gate.py diff-gate --base origin/dev` -> `doc-gate: clean`.
Swept `docs/`, `README*`, `CONTRIBUTING.md`, `AGENTS.md`, `docs/agent-*.md`: the only
SSRF prose is `docs/userspace-app-capabilities.md:297`, which documents
`tinyagentos/userspace/url_guard.py` (a separate install-time guard, untouched here)
and is not made stale by this change. The behaviour contract for this helper lives in
its module docstring, which is rewritten to state the pinning rule and that any client
fetching a user-supplied URL must come from `guarded_async_client`.
Changelog fragment: `changelog.d/tsk-6uymvv-ssrf-dns-pinning.md`.

Note for reviewers: passing an explicit transport means `httpx` no longer reads
`HTTP_PROXY`/`HTTPS_PROXY` from the environment for these fetches. That is
intentional and documented in the factory's docstring - a proxy would do the
resolving and defeat the pin.

## Follow-ups (not in this card's scope)

- `tinyagentos/scheduling/mesh_sync.py:109` `is_safe_url()` is a second, independent
  SSRF implementation with the same validate-then-fetch shape.
- `tinyagentos/userspace/url_guard.py:26` `resolve_safe_public_ip()` is a third; it
  already resolves and returns the address, but its caller does not pin to it.

## Summary by CodeRabbit

* **Security Improvements**
  * Strengthened protection against DNS-rebinding and SSRF attacks by ensuring outbound requests connect only to addresses that passed validation.
  * Applied guarded networking to web fetching, downloads, proxying, extraction, peer handshakes, and push notifications.
  * Preserved TLS hostname verification and redirect security checks.
  * Blocked Unix-socket connections through the SSRF-protected transport.

### CODE REVIEW

File: tinyagentos/library_pipeline.py [hunk 1]
```diff
@@ -491,7 +492,7 @@ async def _fetch() -> tuple[str, str, bytes]:
             for _hop in range(self._MAX_WEB_REDIRECTS + 1):
                 validate_url_or_raise(current_url)
 
-                async with httpx.AsyncClient(
+                async with guarded_async_client(
```
[the remark this answers is not in the document]
Comment (jaylfc) [reply tinyagentos/library_pipeline.py]: Fixed in 50fc288d: hoisted `async with guarded_async_client(...) as client:` outside the `for _hop` redirect loop in `_fetch()` — one client now serves every hop of the redirect chain instead of a fresh pool/SSL-context/pinned-backend per hop. Added `test_web_redirect_reuses_single_guarded_client` in `tests/test_library.py`, which fails on the pre-fix code (entered 3 times for a 3-hop fetch) and passes after. Existing SSRF rebinding tests (`tests/test_ssrf_rebinding.py`) stay green — the per-connection re-resolve/validate still happens via `_PinnedResolutionBackend.connect_tcp`.

File: tinyagentos/routes/desktop_browser/ssrf.py [hunk 2]
```diff
@@ -112,7 +154,10 @@ def validate_url_or_raise(url: str, *, allow_private: bool = False) -> None:
                 results = socket.getaddrinfo(host, None)
                 # results is a list of (family, type, proto, canonname, sockaddr).
                 # sockaddr[0] is the address string for both AF_INET and AF_INET6.
-                addrs = list({r[4][0] for r in results})
+                # Dedupe but keep resolver order: the first answer is the one a
+                # pinned connection uses, and getaddrinfo already sorts by the
+                # RFC 6724 destination preference.
+                addrs = list(dict.fromkeys(r[4][0] for r in results))
```
[the remark this answers is not in the document]
Comment (jaylfc) [reply tinyagentos/routes/desktop_browser/ssrf.py:160]: Fixed in 50fc288d: reworded the comment to drop the RFC 6724 claim. It now says: "Dedupe but keep first-seen order: dict.fromkeys preserves the resolver's own order, and the first answer is the one the pin connects to." Docs-only, no test change.

File: tinyagentos/knowledge_ingest.py [hunk 3]
```diff
@@ -286,27 +294,37 @@ async def _download_article(
...
+        from contextlib import nullcontext
         from urllib.parse import urljoin
 
         from tinyagentos.routes.desktop_browser.ssrf import (
             SsrfBlockedError,
+            guarded_async_client,
             validate_url_or_raise,
         )
 
+        client_cm = (
```
[the remark this answers is not in the document]
Comment (jaylfc) [reply tinyagentos/knowledge_ingest.py:326]: Fixed in 50fc288d: `_download_article` now raises `TypeError("fetch_client must be built by guarded_async_client")` when a caller-supplied `fetch_client` is an `httpx.AsyncClient` whose transport is not `SsrfGuardedAsyncTransport`. Non-`httpx.AsyncClient` test doubles (mocks/fakes) still pass through unchanged, so the existing mock-based tests are untouched. Added `test_fetch_client_must_be_guarded` in `tests/test_knowledge_ingest.py`, which fails pre-fix (`Failed: DID NOT RAISE TypeError`, using a real `httpx.AsyncClient` over `httpx.MockTransport` so no real network is touched) and passes after.

File: tests/test_knowledge_ingest.py [hunk 4]
```diff
@@ -301,6 +305,7 @@ async def test_max_concurrent_zero_raises(store, mock_http):
         IngestPipeline(
             store=store,
             http_client=mock_http,
+        fetch_client=mock_http,
```
[the remark this answers is not in the document]
Comment (jaylfc) [reply tests/test_knowledge_ingest.py]: Fixed in 50fc288d: re-indented `fetch_client=mock_http,` to 12 spaces to match the surrounding `IngestPipeline(...)` call arguments. Style-only.

File: tinyagentos/routes/desktop_browser/ssrf.py [hunk 5]
```diff
@@ -166,6 +213,109 @@ def validate_resolved_addr(addr: str, *, allow_private: bool = False) -> None:
...
+
+    async def connect_tcp(
+        self,
+        host: str,
+        port: int,
+        timeout: float | None = None,
+        local_address: str | None = None,
+        socket_options: typing.Iterable[typing.Any] | None = None,
+    ) -> httpcore.AsyncNetworkStream:
+        addrs = resolve_and_validate(host, allow_private=self._allow_private)
```
[the remark this answers is not in the document]
Comment (jaylfc) [reply tinyagentos/routes/desktop_browser/ssrf.py:243]: Declined for now: the pin's contract is "connect to the address that was validated," not "retry until something answers." Iterating through `addrs[1:]` on a connect failure adds retry/fallback logic to a security-critical path to handle a round-robin-DNS edge case nobody has actually hit. If a real failure of this kind shows up in practice, a follow-up can add it behind a dedicated test at that point.

File: tinyagentos/routes/desktop_browser/ssrf.py [hunk 6]
```diff
@@ -166,6 +213,109 @@ def validate_resolved_addr(addr: str, *, allow_private: bool = False) -> None:
...
+    `httpx.AsyncClient` (timeout, follow_redirects, cookies, headers, ...).
+
+    TLS verification stays on (`verify` defaults to True) and still checks
+    the certificate against the original hostname; pinning happens below
+    the URL, so there is no reason to weaken it. Note that supplying a
+    transport means `httpx` no longer picks up HTTP_PROXY/HTTPS_PROXY from
+    the environment — proxying an untrusted fetch would defeat the pin
+    anyway, since the proxy would do the resolving.
+    """
+    return httpx.AsyncClient(
```
[the remark this answers is not in the document]
Comment (jaylfc) [reply tinyagentos/routes/desktop_browser/ssrf.py:311]: Not applicable: `verify` and `http2` are keyword-only named parameters in `guarded_async_client`'s own signature (ssrf.py:291-297), not positionally consumed out of `**kwargs`. A call like `guarded_async_client(verify=False, ...)` binds `verify` to the named parameter directly; a caller cannot also smuggle `verify` into `**kwargs` since Python raises `TypeError: got multiple values for keyword argument 'verify'` on a repeated keyword before the function body ever runs. There's nothing here to assert or guard against.
